src/v/cluster/jumbo_log_stm.cc

```diff
@@ -0,0 +1,221 @@
+    // intent id allocation invariants.
+    auto r = co_await _raft->replicate(
+      _insync_term,
+      model::make_memory_record_batch_reader(std::move(batch_builder).build()),
+      raft::replicate_options(raft::consistency_level::quorum_ack));
+    if (!r) {
+        throw std::system_error(raft::make_error_code(raft::errc::timeout));
+    }
+
+    // TODO(nv): Sure ok to timeout here?
+    if (!co_await wait_no_throw(
+          model::offset(r.value().last_offset()),
+          model::timeout_clock::now() + timeout)) {
+        throw std::system_error(raft::make_error_code(raft::errc::timeout));
+    }
+
+    co_return intent_id;
+}
+
+ss::future<bool> jumbo_log_stm::sync(model::timeout_clock::duration timeout) {
+    auto is_synced = co_await persisted_stm::sync(timeout);
+
+    co_return is_synced;
+}
+
+ss::future<> jumbo_log_stm::apply(const model::record_batch& batch) {
+    if (batch.header().type != model::record_batch_type::jumbo_log_cmd) {
+        co_return;
+    }
+
+    batch.for_each_record([this](model::record&& r) {
+        auto key = serde::from_iobuf<cmd_key>(r.release_key());
+
+        switch (key) {
+        case create_write_intent_cmd::key: {
+            auto intent = serde::from_iobuf<create_write_intent_cmd::value>(
+              r.release_value());
+
+            apply_create_write_intent(std::move(intent));
+            break;
+        }
+        default:
+            throw std::runtime_error(fmt_with_ctx(
+              fmt::format,
+              "Unknown jumbo_log_stm command {}",
+              static_cast<int>(key)));
+        }
+
+        return ss::stop_iteration::no;
+    });
+
+    co_return;
+}
+
+ss::future<> jumbo_log_stm::write_snapshot() {
+    if (_is_writing_snapshot) {
+        return ss::now();
+    }
+    _is_writing_snapshot = true;
+    // return _raft
+    //   ->write_snapshot(raft::write_snapshot_cfg(0, iobuf()))
+    //   .finally([this] { _is_writing_snapshot = false; });
+    return ss::now();
+}
+
+ss::future<>
+jumbo_log_stm::apply_local_snapshot(raft::stm_snapshot_header, iobuf&&) {
+    return ss::make_exception_future<>(
+      std::logic_error("jumbo_log_stm doesn't support snapshots"));
+}
+
+ss::future<raft::stm_snapshot> jumbo_log_stm::take_local_snapshot() {
+    return ss::make_exception_future<raft::stm_snapshot>(
+      std::logic_error("jumbo_log_stm doesn't support snapshots"));
+}
+
+ss::future<> jumbo_log_stm::apply_raft_snapshot(const iobuf&) {
+    return ss::now();
+}
+
+void jumbo_log_stm::apply_create_write_intent(write_intent intent) {
+    auto intent_id = intent.segment.id;
+    auto object_id = intent.segment.object.id;
+
+    // Using vassert to avoid having to deal with exception safety which can
+    // violate consistency invariants between data structures.
+
+    vassert(
+      intent_id == _next_write_intent_id,
+      "Invariant violation: Write intent ID {} does not match expected ID {}",
+      intent_id,
+      _next_write_intent_id);
```

Comment:
maybe it's feasible to remove this by forcing the callers to create a write intent to through an interface that implicitly enforces these rules. but just maybe. i haven't read through all the code yet.